x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyLocks.java

```diff
@@ -0,0 +1,71 @@
+package org.elasticsearch.xpack.enrich;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+
+class EnrichPolicyLocks {
+
+    private final ReadWriteLock coordinationLock = new ReentrantReadWriteLock(true);
+    private final ConcurrentHashMap<String, Semaphore> policyLocks = new ConcurrentHashMap<>();
+    private final AtomicLong policyRunCounter = new AtomicLong(0L);
+
+    static class LockState {
+        final boolean runningPolicies;
+        final long revision;
+
+        LockState(boolean safe, long revision) {
+            this.runningPolicies = safe;
+            this.revision = revision;
+        }
+    }
+
+    void lockPolicy(String policyName) {
+        coordinationLock.readLock().lock();
+        try {
+            Semaphore runLock = policyLocks.computeIfAbsent(policyName, (name) -> new Semaphore(1));
+            boolean acquired = runLock.tryAcquire();
+            if (acquired == false) {
+                throw new EsRejectedExecutionException("Policy execution failed. Policy execution for [" + policyName +
+                    "] is already in progress.");
+            }
+            policyRunCounter.incrementAndGet();
+        } finally {
+            coordinationLock.readLock().unlock();
+        }
+    }
+
+    LockState lockState() {
+        if (coordinationLock.writeLock().tryLock()) {
+            try {
+                long revision = policyRunCounter.get();
+                int currentPolicyExecutions = policyLocks.size();
```

Comment:
nit: prefer mappingCount over size for CHM (however, it doesn't make a different here)